#451 implemented topic selection service for graph page

Mardan21 opened the pull request:
## Changes

Implemented a centralized `TopicSelectionService` to manage topic selection state across the graph page components. This replaces the previous approach of passing callback functions between components with a reactive service pattern.

- Created `TopicSelectionService` with reactive state management using BehaviorSubject
- Updated graph-page component to subscribe to service changes instead of using direct function calls
- Updated desktop and mobile sidebar components to use service methods (`add`, `remove`, `clear`)
- Updated node-display component to use service for topic selection
- Removed legacy input properties and function callbacks from all sidebar components
- Maintained existing UI behavior, animations, and user experience

## Notes

_Any other notes go here_

## Test Cases

- Desktop topic selection via tree component
- Mobile topic selection via node display cards
- Graph updates when topics are selected/deselected
- Clear all topics functionality
- Service state synchronization across components
- No console errors or TypeScript compilation issues
- Smooth animations and UI interactions maintained

## Checklist

It can be helpful to check the `Checks` and `Files changed` tabs.
Please review the [contributor guide](https://nerdocs.atlassian.net/wiki/spaces/NER/pages/8060929/Software+Contributor+Guide) and reach out to your Tech Lead if anything is unclear.
Please request reviewers and ping on slack only after you've gone through this whole checklist.

- [x] All commits are tagged with the ticket number
- [x] No linting errors / newline at end of file warnings
- [x] All code follows repository-configured prettier formatting
- [x] No merge conflicts
- [x] All checks passing
- [ ] Screenshots of UI changes (see Screenshots section)
- [x] Remove any non-applicable sections of this template
- [x] Assign the PR to yourself
- [x] No `package-lock.json` changes (unless dependencies have changed)
- [x] Request reviewers & ping on Slack
- [x] PR is linked to the ticket (fill in the closes line below)

Closes #451

Mardan21 committed:
feat: implement topic selection service for graph page

- Add TopicSelectionService with add/remove/set/clear methods
- Update graph-page component to use service instead of direct function passing
- Update desktop and mobile sidebar components to use service
- Update node-display component to use service for topic selection
- Remove legacy input properties and function callbacks
- Maintain existing UI behavior and animations

bracyw reviewed (changes requested):
Graph screen crashes.

Mardan21 committed:
feat: Implement centralized topic selection service (#451)

Mardan21 commented:
Fixed the issue causing the graph screen crash. The service subscription was being cleared when topics changed - separated persistent vs. data subscriptions to maintain reactivity. Tested topic selection/deselection and historical data loading, all working for me.

Mardan21 committed:
fix: prevent infinite loading by using persistent subscriptions for essential queries

Mardan21 committed:
fix: update method calls to match TopicSelectionService API

bracyw commented on angular-client/src/pages/graph-page/graph-page.component.ts:
```diff
@@ -432,10 +459,21 @@ export default class GraphPageComponent implements OnInit, OnDestroy {
+   * This updates the shared service state, which will trigger the subscription
+   * in ngOnInit to process the selection change.
+   * @param dataTypes Array of data types to set as selected
    */
```
Is this still needed, you may be able to delete.

bracyw commented on angular-client/src/pages/graph-page/graph-sidebar/graph-sidebar-desktop/graph-sidebar-desktop.component.ts:
```diff
@@ -80,7 +80,13 @@ export default class GraphSidebarDesktopComponent implements OnInit, OnDestroy {
+          //   // Mark all parents as containing selected nodes
+          //   parents.forEach((parent) => containsSelectedNode.set(parent, true));
+          // }
+          if (node.selectable && node.data?.dataType && this.topicSelectionService.isSelected(node.data.dataType)) {
```
This is so much nicer!

bracyw commented on angular-client/src/pages/graph-page/graph-sidebar/graph-sidebar-desktop/graph-sidebar-desktop.component.ts:
```diff
@@ -80,7 +80,13 @@ export default class GraphSidebarDesktopComponent implements OnInit, OnDestroy {
         for (const node of nodes) {
           // Check if this is a leaf node and matches a selected data type
-          if (node.selectable && node.data?.dataType && this.currentSelectedDataTypes().includes(node.data.dataType)) {
+          // if (node.selectable && node.data?.dataType && this.currentSelectedDataTypes().includes(node.data.dataType)) {
```
delete

bracyw commented on angular-client/src/pages/graph-page/graph-sidebar/graph-sidebar-desktop/graph-sidebar-desktop.component.ts:
```diff
@@ -114,26 +120,40 @@ export default class GraphSidebarDesktopComponent implements OnInit, OnDestroy {
   }
 
+  // clearSelections = () => {
+  //   this.treeNodes.forEach((node) => {
```
delete

bracyw commented on angular-client/src/pages/graph-page/graph-sidebar/graph-sidebar-desktop/graph-sidebar-desktop.component.ts:
```diff
@@ -114,26 +120,40 @@ export default class GraphSidebarDesktopComponent implements OnInit, OnDestroy {
     return dataTypeNamePipe(dataTypeName);
   }
 
+  // nodeSelect(event: TreeNodeSelectEvent) {
```
delete

bracyw commented on angular-client/src/pages/graph-page/graph-sidebar/graph-sidebar-desktop/graph-sidebar-desktop.component.ts:
```diff
@@ -148,15 +168,22 @@ export default class GraphSidebarDesktopComponent implements OnInit, OnDestroy {
     /* leaf rows fall through → normal multi‑selection behaviour */
   }
 
+  // onNodeUnselect(event: TreeNodeUnSelectEvent) {
```
delete

bracyw commented on angular-client/src/pages/graph-page/graph-sidebar/graph-sidebar.component.html:
```diff
@@ -1,11 +1,7 @@
+    <graph-sidebar-mobile [dataTypes]="this.dataTypes()" />
   </div>
 } @else {
-  <graph-sidebar-desktop
```
So clean!

bracyw reviewed (changes requested):
Some really quick changes. This is going to be so helpful! Also if you can run formatter that would be awesome.

Mardan21 committed:
refactor: clean up graph-page components and remove dead code

- Remove unused setSelectedDataTypes wrapper method
- Remove commented-out old implementations in sidebar-desktop
- Clean up graph-sidebar template (remove unnecessary div and this. prefix)

Mardan21 committed:
formatting files with prettier

bracyw reviewed (approved):
LGTM!
